### ISSUE TITLE
Understand current performance bottlenecks in signature generation

### ISSUE THREAD

**Issue by jakmeier** [sig-net/mpc#72]
### Description

Currently, when the dev network starts and generates triples and pre-signatures at full speed, we only observe around 40 messages per second on each node. ([see graph and comment here](https://github.com/sig-net/mpc/issues/52#issuecomment-2525794265))

We need to understand what the limiting factors are.

In a perfect world, we are only limited by the CPU-time it takes to perform the cryptographic work and the network delay. We can achieve this if we ensure 3 things.

1. Incoming messages are always immediately delivered to [cait-sith](https://github.com/cronokirby/cait-sith) through [Protocol::message](https://docs.rs/cait-sith/latest/cait_sith/protocol/trait.Protocol.html#tymethod.message).
2. We are always calling [Protocol::poke](https://docs.rs/cait-sith/latest/cait_sith/protocol/trait.Protocol.html#tymethod.poke) until it tells us to wait.
3. Any messages generated are immediately sent.

In theory, these 3 tasks can always run in parallel, without blocking each other. Anytime one of these three tasks is not running, we potentially introduce overhead that could be avoided.

In practice, `Protocol::poke` and `Protocol::message` both require mutable access to `Protocol`, so they will not be able to run in parallel on the same Protocol instance. But presumably `Protocol::message` doesn't do actual work and only records the incoming message (to be checked), so the overhead of doing it serially should be minimal.

In any case, we should try to find out if any of these three tasks is delayed significantly due to implementation inefficiencies.

### Possible steps

1. We can add specific metrics to help us understand how much time we spend on each task. (Created as a separate issue: https://github.com/sig-net/mpc/issues/71)
2. Look at general execution traces from [`tracing`](https://lib.rs/crates/tracing) to see if anything looks suspicious. See [here](https://near.github.io/nearcore/practices/workflows/otel_traces.html) for how it's done in nearcore.
    - Note: I see we have some tracing tooling already in the code base but I am not sure how much it is used and how well maintained. It might be worth it to invest some time into this by adding appropriate spans and setting up good tooling to look at the timing of the traces.
3. Observe machines while they are under load with general tools like [htop(1)](https://www.man7.org/linux/man-pages/man1/htop.1.html), [iotop(8)](https://www.man7.org/linux/man-pages/man8/iotop.8.html), [perf(1)](https://www.man7.org/linux/man-pages/man1/perf.1.html), or more specialized tools like [tokio-console](https://github.com/tokio-rs/console) if we can compile the nodes with nightly tokio features.

**Comment by jakmeier** [comment]
If anyone already has data to help understand the performance bottlenecks better, or knows a good way to find it out, please share it here. :)

**Comment by jakmeier** [comment]
Related issue with relevant analysis: https://github.com/sig-net/mpc/issues/32

**Comment by volovyks** [comment]
Are you suggesting replacing Prometheus with Tracing? Or built these new metrics using it?
It looks promising, but it will require refactoring.

**Comment by jakmeier** [comment]
No, I would still keep Prometheus for old and new metrics.

Tracing is an additional tool for deeper performance investigations. It can work more generally, including places where we haven't added metrics. And it can potentially give more fine-grained information, telling you exactly how many micro seconds are spend on each function. But it might require you to add more tracing information at runtime to be useful.

I see today logs are akready done with the tracing macros (e.g. `tracing::warn!`) and we even have some span info added with macros, too (e.g. `tracing::info_span!` and `#[tracing::instrument]`). This means, you are already producing at least some tracing data.

This data is then consumed by a tracing subscriber. [This code](https://github.com/sig-net/mpc/blob/2bc9f78fbfc563ba65c7911033a5504a0d9e91b7/chain-signatures/node/src/cli.rs#L164-L172) suggest you have this integrated with Google's Stackdriver. I'm not familiar with Stackdriver but perhaps that's all you need to look at the execution traces in detail and you can use it already today for investigating performance issues. Maybe you can find out which functions a node spends most time in.

At nearcore, [Jaeger](https://www.jaegertracing.io/) is used for presenting the traces. The tokio documentation chapter [Next steps with Tracing](https://tokio.rs/tokio/topics/tracing-next-steps) has an example how to set this up. But I wouldn't add something new before understanding what you already have (Stackdriver).

**Comment by volovyks** [comment]
Google provides [Google Cloud Profiler](https://cloud.google.com/profiler/docs/measure-app-performance) and [Tracing](https://cloud.google.com/trace/docs/) functionality. However, it is not turned on for our project, and it does not appear to support Rust projects natively. But I understand what you mean. Profiling and its flame graphs should give us many insights.
@auto-mausx have you worked with it? Have you seen it working for Rust? I have not found much information about it.

**Comment by auto-mausx** [comment]
So as with any tracing profiler, we will need to send traces that make the visualization useful. I have used GCP Tracing before, albeit just as a simple functionality check. We can utilize Google or grafana as they both have Tracing support via OpenTelemetry.

**Comment by volovyks** [comment]
stale
